txscript/sigcache_test.go

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2015 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package txscript
+
+import (
+	"crypto/rand"
+	"testing"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/btcsuite/btcd/wire"
+)
+
+// genRandomSig returns a random message, public key, and a signature of the
+// message under the public key. This function is used to generate randomized
+// test data.
+func genRandomSig() ([]byte, *btcec.Signature, *btcec.PublicKey, error) {
+	privKey, err := btcec.NewPrivateKey(btcec.S256())
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	msgHash := make([]byte, wire.HashSize)
+	if _, err := rand.Read(msgHash); err != nil {
+		return nil, nil, nil, err
+	}
+
+	sig, err := privKey.Sign(msgHash)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	return msgHash, sig, privKey.PubKey(), nil
+}
+
+// TestSigCacheAdd tests the ability to add, and later check the existence of a
+// signature triplet in the signature cache.
+func TestSigCacheAddExists(t *testing.T) {
+	sigCache := NewSigCache(200)
+
+	// Generate a random sigCache entry triplet.
+	msg1, sig1, key1, err := genRandomSig()
+	if err != nil {
+		t.Errorf("unable to generate random signature test data")
+	}
+
+	// Add the triplet to the signature cache.
+	sigCache.Add(msg1, sig1, key1)
+
+	// The previously added triplet should now be found within the sigcache.
+	if !sigCache.Exists(msg1, sig1, key1) {
```

Comment:
As I partially showed in another comment about why the `sigInfo` struct won't work properly, this test really should be using deep copies of the message, signature, and key (can be obtained by serializing and deserializing them) to make sure that you're actually checking for existence of the actual message/sig/key versus the memory they point to.